### PR TITLE
Allow option to opt-out of provided memory allocator

### LINKED ISSUE #1414: Consider add dlmalloc-rs as allocator or opt-out the BumpAllocator
Since `wee_alloc` has been removed in https://github.com/paritytech/ink/pull/1403, BumpAllocator is the only choice now.
However, BumpAllocator is not suitable for our off-chain execution env, because it doesn't free any memory at all.

There are two options to resolve it in my mind:
- Support [dlmalloc-rs](https://github.com/alexcrichton/dlmalloc-rs), which is the allocator for wasm in std, as an opt-in replacement to the bump allocator.
- Add a feature `no_allocator` to crate `ink_allocator` to allow downstream to opt-out of the bump allocator.

### PR DESCRIPTION
**Pull request by HCastano**
This takes https://github.com/paritytech/ink/pull/1415 one step further and provides the
`no-allocator` feature to the top level `ink` crate, allowing contract authors to opt-out
of the provided memory allocator.

It also adds an example showing how a contract author could provide their own allocator.

Closes #1414.

cc @kvinwang

### CODE REVIEW

File: examples/custom_allocator/lib.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
+//! # Custom Allocator
+//!
+//! This example demonstrates how to opt-out of the ink! provided global memory allocator.
```
Comment (cmichi) [inline examples/custom_allocator/lib.rs:3]: Lemme be the guy commenting about something on the order of 80 CPL 😬 .

File: examples/custom_allocator/lib.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+//! If you don't handle allocations correctly you can introduce security vulnerabilities to your
+//! contracts.
+//!
+//! You may also introduce performance issues. This is because the code of your allocator will
+//! be included in the final contract binary, potentially increasing gas usage significantly.
+//!
+//! ## Why Change the Allocator?
+//!
+//! The default memory allocator was designed to have a tiny size footprint, and made some
+//! compromises to achieve that, e.g it does not free/deallocate memory
```
Comment (cmichi) [inline examples/custom_allocator/lib.rs:21]:
```suggestion
//! compromises to achieve that, e.g it does not free/deallocate memory.
```

**Review by cmichi — APPROVED** [review]
LGTM. Just the nitpicks about having less CPL throughout.

File: examples/custom_allocator/lib.rs (hunk 1 above)
Comment (ascjones) [reply examples/custom_allocator/lib.rs:3]: Can we configure `rustfmt` to catch this? I think it enforces code CPL to 100, not sure about the comments.

**Comment by cmichi** [comment]
Add the example to https://github.com/paritytech/ink-examples as well, please.

File: examples/custom_allocator/lib.rs (hunk 1 above)
Comment (HCastano) [reply examples/custom_allocator/lib.rs:3]: We already enforce this to 90 characters, which is what it's at 😉 

https://github.com/paritytech/ink/blob/b8862a1147a8fb577c82304ed1dec9a8c45c386a/.rustfmt.toml#L1

**Pull request merged**